src/Features/CSharp/Portable/Snippets/CSharpForLoopSnippetProvider.cs

```diff
@@ -0,0 +1,179 @@
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp.CodeStyle;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Snippets;
+using Microsoft.CodeAnalysis.Snippets.SnippetProviders;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+using static Humanizer.In;
+
+namespace Microsoft.CodeAnalysis.CSharp.Snippets
+{
+    [ExportSnippetProvider(nameof(ISnippetProvider), LanguageNames.CSharp), Shared]
+    internal class CSharpForLoopSnippetProvider : AbstractForLoopSnippetProvider
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpForLoopSnippetProvider()
+        {
+        }
+
+        /// <summary>
+        /// Creates the for loop statement syntax.
+        /// Must be done in language specific file since there is no generic way to generate the syntax.
+        /// </summary>
+        protected override async Task<SyntaxNode> CreateForLoopStatementSyntaxAsync(Document document, CancellationToken cancellationToken)
+        {
+            var compilation = await document.Project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+            var generator = SyntaxGenerator.GetGenerator(document);
+            var options = await document.GetAnalyzerConfigOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var varBuiltInType = false;
+
+            if (options != null)
+            {
+                varBuiltInType = options.GetOption(CSharpCodeStyleOptions.VarForBuiltInTypes).Value;
+            }
+
+            var indexVariable = generator.Identifier("i");
+            var varIdentifier = SyntaxFactory.IdentifierName("var");
+
+            // Creating the variable declaration based on if the user has
+            // 'var for built in types' set in their editorconfig.
+            var variableDeclarationSyntax = varBuiltInType
+                ? SyntaxFactory.VariableDeclaration(varIdentifier,
+                    SyntaxFactory.SingletonSeparatedList(
+                        SyntaxFactory.VariableDeclarator(
+                            indexVariable,
+                            argumentList: null,
+                            SyntaxFactory.EqualsValueClause((ExpressionSyntax)generator.LiteralExpression(0)))))
+                : SyntaxFactory.VariableDeclaration(compilation.GetSpecialType(SpecialType.System_Int32).GenerateTypeSyntax(allowVar: false),
```

Comment:
@davidwengier To my knowledge, it adds `DoNotAllowVarAnnotation.Annotation`. This annotation is used by `CSharpUseImplicitTypeHelper`.

https://github.com/dotnet/roslyn/blob/a84cae03689143e95ffd5a26efa8ef010e5c1f3f/src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.cs#L60-L61

https://github.com/dotnet/roslyn/blob/a84cae03689143e95ffd5a26efa8ef010e5c1f3f/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Utilities/TypeStyle/CSharpUseImplicitTypeHelper.cs#L38-L41

One important use of `CSharpUseImplicitTypeHelper` is `CSharpVarReducer`, which is run after every `CodeAction` is applied.

I think no code actions are involved here, so most likely this annotation will do nothing?

I'm curious though why we will ever want to put a `DoNotAllowVarAnnotation.Annotation`

Reply:
@davidwengier Note that if this was a `CodeAction` (e.g, in a CodeFixProvider or CodeRefactoringProvider), we can **always** generate a non-var TypeSyntax and add `Simplifier.Annotation`, and it will magically get converted to var if user prefers it.

See https://github.com/dotnet/roslyn-analyzers/pull/6023 :)

Reply:
Ooh snippets are calling `CleanupDocumentAsync`:

https://github.com/dotnet/roslyn/blob/a84cae03689143e95ffd5a26efa8ef010e5c1f3f/src/Features/Core/Portable/Snippets/SnippetProviders/AbstractSnippetProvider.cs#L106

So things I was saying specific to `CodeAction`s might be applicable to snippets as well.

@akhera99 Can you try to unconditionally produce `compilation.GetSpecialType(SpecialType.System_Int32).GenerateTypeSyntax(allowVar: false).WithAdditionalAnnotations(Simplifier.Annotation)`?

The `CleanupDocumentAsync` call should be able to convert to `var` if user prefers that.

Reply:
@Youssef1313 yep you were correct, thanks :)